Fix #389: Allow secondary alphabetical sorting

randallreedjr opened the pull request:
- After sorting imports by group, allow sorting alphabetically within a group
- import/order rule now accepts sort-order option that may be 'ignore' (default) or 'alphabetical'
- Fixes #389

randallreedjr commented:
This is my first attempt at trying to implement a secondary alphabetical sort on imports. I would appreciate any feedback!

randallreedjr commented:
I resolved the test coverage decrease. Not sure why appveyor is failing, but it appears to be happening when trying to run coveralls.

jfmengels commented on src/rules/order.js:
```diff
@@ -205,6 +269,7 @@ module.exports = {
       },
       'Program:exit': function reportAndReset() {
         makeOutOfOrderReport(context, imported)
+        makeOutOfAlphabeticalOrderReport(context, imported, sortOrder)
```
Another separate thing:
I think it would be nice to to not re-report the modules that were reported in `makeOutOfOrderReport`.

Let's assume lowercase to be part of group 1, uppercase of group 2, and group 1 should appear before group 2.
```js
import 'B';
import 'A';
import 'b';
import 'a';
```
I haven't tried it yet, but I wouldn't be surprised if there were two error reports for moving B to a latter location or something. I would love some tests on that too, though the problem might not be shown with this example.

The reason I'm worried about this, is when you introduce the `order` rule to a new codebase, where imports are sorted in random ways, you'll get a lot of errors for the import section. I'd like for it to be the least confusing people and avoid having 20 errors when you required 10 modules. Note that the code might actually do this already, but without tests I have trouble being sure.

By not re-reporting modules, I think `findOutOfAlphabeticalOrder` could be simplified a bit too. I'm a bit confused about its inner workings at the moment.

Also, with only the `makeOutOfOrderReport` report, you have a consistent `should occur before`/`should occur after` message. I think it would be nice if that was true for all of the reports. Maybe compile the whole list with the group and the sort options one way, then do it the other way and take the shorter one.

Sorry to add a lot of work :sweat_smile:

jfmengels commented on src/rules/order.js:
```diff
@@ -17,6 +17,16 @@ function reverse(array) {
 
+function alphabetical(firstName, secondName, direction) {
+  if (direction === 'a-z') {
+    return firstName < secondName
```
The same module could be imported twice, so this should be `<=` and `>=`. Please add a test for that case :)

jfmengels commented on src/rules/order.js:
```diff
@@ -17,6 +17,16 @@ function reverse(array) {
   }).reverse()
 }
 
+function alphabetical(firstName, secondName, direction) {
```
Name it something like `compareAlphabetically`? Or `isBeforeAlphabetically` (berk). They're ugly names, so if you have better ideas, go for it, but at the moment, without looking at the code and by just looking at places where it's called, it's hard to know what exactly this function does.

randallreedjr replied:
`isAlphabetized`?

jfmengels commented on src/rules/order.js:
```diff
@@ -31,6 +41,31 @@ function findOutOfOrder(imported) {
+      direction
+    )
+    const reversedDirection = direction.split('').reverse().join('')
+    if (alphabetical(importedModule.name, maxSeenAlphabeticalNode.name, reversedDirection)) {
```
Unless the two names are the same, isn't this just equal to `resAlphabetical`?

randallreedjr commented on docs/rules/order.md:
````diff
@@ -141,6 +141,42 @@ import index from './';
 import sibling from './foo';
 ```
 
+### `sort-order: [ignore|alphabetical]`:
````
Sure

randallreedjr commented:
@jfmengels I reworked the approach to solving this problem, please have a look and let me know what you think!

randallreedjr commented:
Anything I can do to help with the review of this PR?

ctavan commented:
This is definitely an addition I've been looking for! It would be great if it could get merged at some point.

@randallreedjr just one remark: I suppose the `yarn.lock` was commited by accident, wasn't it?

randallreedjr commented:
@ctavan Good catch, thanks!

randallreedjr commented:
@benmosher @jfmengels This PR has been open for 7 weeks now. Is there anything I can do to help get it merged in?

mrtnbroder commented:
@jfmengels it would be nice to get this in asap!

randallreedjr commented:
@jfmengels @benmosher Is it time to rebase and resubmit this PR again?

jfmengels commented on docs/rules/order.md:
````diff
@@ -162,6 +162,42 @@ import index from './';
+
+```js
+/* eslint import/order: ["error", {"sort": "ignore"}] */
+/* eslint import/order: ["error"] */
````
This line should be removed

randallreedjr replied:
👍

jfmengels commented on tests/src/rules/order.js:
```diff
@@ -482,7 +523,7 @@ ruleTester.run('order', rule, {
       }, {
         ruleId: 'order',
-        message: '`fs` import should occur before import of `./sibling`',
+        message: '`fs` import should occur before import of `async`',
```
The error should stay as it was before ideally :thinking:

randallreedjr replied:
So the original order is:
```
sibling
async
fs
```
And I believe the desired order is
```
fs
async
sibling
```
So saying that `fs` should be before `./sibling` is technically incorrect, as it leaves out the additional requirement that `fs` also appear before `async`. 

Do you agree?

jfmengels commented on src/rules/order.js:
```diff
@@ -212,7 +230,9 @@ module.exports = {
       'Program:exit': function reportAndReset() {
-        makeOutOfOrderReport(context, imported)
+        const alphabetize = (options['sort'] === 'alphabetical')
+        let [forwardSortComparator, reverseSortComparator] = determineComparators(alphabetize)
```
Make this a const

randallreedjr replied:
👍

randallreedjr commented:
@jfmengels Thanks so much for your comments, I'll get right on it!

randallreedjr commented:
@jfmengels re invalid tests with `sort: ignore`, I will add some

randallreedjr commented:
@jfmengels I ran into a snag with ignoring uppercase. When the filename is capitalized, it is not found in the built-in modules - so for example, `Http` receives a different rank than `http`. This makes the alphabetical sorting irrelevant.
If we wanted to ignore case in the alphabetical sort, I think we'd also need to update src/core/importType to ignore case.
Here's the pertinent code: https://github.com/benmosher/eslint-plugin-import/blob/b9397183562bb38192b9562dfe5284ee3e89ccb1/src/core/importType.js#L17

randallreedjr commented:
@jfmengels I just pushed my latest branch. I think I addressed all of your comments, including
* Adding a new invalid test with sort: ignore (and incorrect groups)
* Ignoring capitalization for non-builtin modules
* Adding tests for interactions between the alphabetical sort and newlines-between rules
* Deleting the extra comment line in order.md
* Changing `let` to `const` for the comparators declaration

Last outstanding point is the change to the rule for message:
```
'`fs` import should occur before import of `./sibling`' 
```
(see my reply above)

Let me know when you've had a chance to review! Thanks!

randallreedjr commented:
@jfmengels Just a reminder that this pull request is still awaiting review. Please let me know if further changes are necessary.

rjhilgefort commented:
@jfmengels Might you have some time to review the updates and provide additional feedback (or a timeframe on when you might have time to do so)? 

I would love to see this get pushed through as it's been a pain point for me as well. Thanks for the work on this plugin and thanks @randallreedjr for the PR!

rmacklin commented:
Thanks for working on this @randallreedjr. I can't wait to use this to have eslint automatically enforce alphabetical imports within groups, so I can stop manually commenting on it during reviews!

ismay commented:
Yeah awesome work @randallreedjr! Hope this'll be accepted soon

nunofgs commented:
Yes please. I'm waiting for this PR as well.

randolphpark commented:
Yes please. I'm waiting for this PR as well.

ismay commented:
@jfmengels Seems like @randallreedjr has addressed your comments. I'm not sure if it's a problem that the appveyor build is failing? Is there anything we can do to get this merged?

randolphpark commented:
@jfmengels Seems like @randallreedjr has addressed your comments. Is there anything we can do to get this merged?

mehcode commented:
@benmosher Can we get this merged in? This would be great!

nevir commented:
Any luck with this PR?

jeffcarbs commented:
Would love to see this get merged 👍

dlecocq commented:
If the concern was over the autofix behavior, maybe the rule could get merged now, and the autofix added later?

AndrewBestbier commented:
Adding myself to the list of people who would love to see this merged!

ericyliu commented:
would love to see this as well :+1:

nevir commented:
@ljharb @jfmengels 

I'd be interested in potentially taking this PR over and pushing it through to the finish line - what changes are left in your eyes?

So far, it looks like:

> Could you add some tests to ensure that sorting is not done across non-import non-require code boundaries?

ljharb commented:
@nevir let's start with that :-) please don't open a new PR; instead, post a link here to your branch on your fork, and I'll cherry-pick your commits over into this PR.

nevir commented:
@ljharb I'd like to push back on that, actually: the order rule's current behavior is to force groups to be combined (see failing tests at https://github.com/nevir/eslint-plugin-import/commit/fc3b4ba93e32b62c23f79a711d824283a91c47d3):

```js
import b from 'b';

// ^ There should be no empty line within import group
console.log('a thing');

import a from 'a';
```

And, honestly, that's good behavior - especially with `import`s being evaluated before all other statements, per the spec.  Will reduce confusion surrounding behavior for people once they're using native imports.

---

Also, I have pushed a continuation of this PR at https://github.com/nevir/eslint-plugin-import/tree/nevir/import-sort-order which syncs it up with master, and fixes the remaining test failures

ljharb commented:
@nevir the actual import and evaluation of it is indeed hoisted to the top; however, the *usage of the binding* is not available (per a temporal dead zone) until the position of the `import`.

nevir commented:
Aha - did not know that detail!  Still, if supporting `require`/`import`s interleaved within the executable code is something we want to support, I think that should be a different PR, since that's broken today.  (It's also a significant refactor to the rule, from what I can tell)

JAStanton commented:
@ljharb any movement on @nevir's comment / this PR?  My team has been wanting this one real bad :) happy to help in anyway we can but it looks like this is blocked by a design decision?

ljharb commented:
> Still, if supporting require/imports interleaved within the executable code is something we want to support

It's something the language and the ecosystem supports, so if this rule doesn't support it as-is, then it needs to be fixed before any functional changes are made to it.

In other words, it's fine if that's a separate PR - but then this PR would be blocked on that one.

nevir commented:
> It's something the language and the ecosystem supports, so if this rule doesn't support it as-is, then it needs to be fixed before any functional changes are made to it.

Doesn't that imply that the current implementation of the `order` rule is busted?

I guess I'm a bit disappointed—pragmatically, that rule does well enough for the vast majority of cases out there.  I can understand the motivation behind not maintaining more code on top of the busted behavior, but it's also quite challenging to support the correct behavior (ignoring this PR), too.  

It's looking like it'll lead to a lot more complexity in the `order` rule, too (requires an understanding of which AST nodes are allowable at the top level & cause groups to be split apart, vs those that do not affect grouping).  Do you have guidance for accomplishing that in a way that's not super brittle (e.g. not a whitelist/blacklist of many, many AST node types)?

benmosher commented:
@nevir: does your continuation have breaking changes from the current state of this rule? just want to know if merging it will be a semver-major event.

benmosher commented:
context: I would love to get this in, but need to manage properly if it's semver-major

AndrewBestbier commented:
Just saw this today: https://blogs.msdn.microsoft.com/typescript/2018/03/27/announcing-typescript-2-8/#organize-imports. Organising imports with TypeScript

nevir commented:
@benmosher it's a Y bump as far as I can tell (should be backwards compatible).  

However, it may need some more work due to #908 (the autofixer will need to be updated to support, or have to be disabled if groups are enabled, I think)

benmosher commented:
makes sense. is this the right PR or is there some way to sync this one with your current work? or is another one open? (sorry, there is a lot in this thread and I haven't had bandwidth to keep up 😞)

nevir commented:
@benmosher https://github.com/nevir/eslint-plugin-import/tree/nevir/import-sort-order is the latest & greatest (LMK if you'd prefer I spin up a different PR)

benmosher commented:
I think I have sufficient authorization to merge your commits to this PR, I'll take a stab at that when I have a chance and will take a look at the conflicts 👍  thanks!

nevir commented:
Awesome, thanks for picking up the torch on this!

snoob commented:
Any news of this PR ? @randallreedjr

randallreedjr commented:
Sorry @snoob, I'm not really sure. I opened this PR over a year and a half ago, and couldn't get any traction with it. At this point, I have no idea what needs to be done to get it merged in.

snoob commented:
@randallreedjr Thank you for your answer :) One last question is the sort working on your branch ? or have you find any workaround ? I'am fed up with sorting manually..

ljharb commented:
The PR is still waiting on https://github.com/benmosher/eslint-plugin-import/pull/629#issuecomment-356100725, as far as i can tell.

ljharb commented:
@snoob please post a link to your branch and ill pull the commits into this pr (as opposed to making a new pr)

snoob commented:
@ljharb : Otherwise i was thinking about committing on @randallreedjr repo via PR ?

ljharb commented:
@snoob that's totally fine too!

snoob commented:
@ljharb too late :p i have already rebased the @randallreedjr on my repo branch https://github.com/snoob/eslint-plugin-import/tree/import-sort-order

ljharb commented:
@snoob thanks for the help, i've used your rebase to update this PR.

janpaul123 commented:
@ljharb cool, makes sense. I think it is good to fix problems with this rule before adding more functionality to it, thanks for reiterating that. 👍 

I unfortunately really needed this functionality though, and also the autofixing (which this PR doesn't include), so for anyone who comes across this thread and needs that too, I made a small package based on [another PR](https://github.com/benmosher/eslint-plugin-import/pull/1105) over here: https://github.com/janpaul123/eslint-plugin-import-order-alphabetical It's not perfect but it suited my needs well enough. 🤷‍♂️

snoob commented:
@ljharb : What do you think about @janpaul123 work ? 

Because we have to fix test on this branch and we don't have autofix also. All seem to pass on @janpaul123 PR. So what do we do ?

ljharb commented:
@snoob any time you have commits you want me to pull into this PR, please ping me here and I'll do so.

cdreve commented:
Hi guys,
Any updates on this PR? I would really like to use this feature.
Thank you.

@janpaul123 @ljharb @snoob @randallreedjr

itajaja commented:
hi @ljharb, I'll be happy to pick up the torch from where it was left. by following the conversation, it seems like the main problem is interleaved code between import groups, but as far as I can tell, that seems resolved. is it something in here specifically that adds a regression? happy to open a new PR for the additional work, or for you to cherry pick from my branch.

ljharb commented:
@itajaja posting a link here would be preferable to opening a new PR; indeed, if the interleaving concerns mentioned upthread are covered by tests, this PR, rebased, might be sufficient.

It's a bit tricky right now, though, because tests are failing on master.

jeffshaver commented:
@ljharb also interested in this. let me know if any help is needed getting this done

VincentLanglet commented:
I think https://github.com/benmosher/eslint-plugin-import/pull/1105 will do the same job

Glinkis commented:
Has there been any update on this? I would love if this was thing.

rjhilgefort commented:
Whoa! It happened! Pumped to be getting this feature in everyone that contributed, thanks!

randallreedjr commented:
Wow, I can't believe it! This was one of my earliest open source pull requests. Thanks @ljharb for merging and everyone else for helping to get this in!!!

techsin commented:
Just to be sure, this only sorts based on path of import NOT on variable being imported.. So `import { prop }..., import Abc..., import *... , etc` will have no effect.
